src/passes/TypeSSA.cpp

```diff
@@ -0,0 +1,257 @@
+// that we write either 5 or 11 into the value of type $A. But after TypeSSA we
+// have this:
+//
+//  x = struct.new $A.x (5);
+//  print(x.value);
+//
+//  y = struct.new $A.y (11);
+//  print(y.value);
+//
+// Now each place has its own type, and a single value is possible in each type,
+// allowing us to infer what is printed.
+//
+// Note that the metaphor of SSA is not perfect here as we do not handle merges,
+// that is, when there is a value that can read from more than one struct.new
+// then we do nothing atm. We could create a phi there, but in general that
+// would require multiple inheritance. TODO think more on that
+//
+
+#include "ir/find_all.h"
+#include "ir/module-utils.h"
+#include "ir/names.h"
+#include "ir/possible-constant.h"
+#include "pass.h"
+#include "wasm-builder.h"
+#include "wasm.h"
+
+namespace wasm {
+
+namespace {
+
+struct News {
+  std::vector<StructNew*> structNews;
+  // TODO: arrayNews of all kinds
+};
+
+struct NewFinder : public PostWalker<NewFinder> {
+  News news;
+
+  void visitStructNew(StructNew* curr) { news.structNews.push_back(curr); }
+
+  // TODO arrays
+};
+
+struct TypeSSA : public Pass {
+  // Only modifies struct/array.new types.
+  bool requiresNonNullableLocalFixups() override { return false; }
+
+  Module* module;
+
+  void run(Module* module_) override {
+    module = module_;
+
+    if (!module->features.hasGC()) {
+      return;
+    }
+
+    // First, find all the struct/array.news.
+
+    ModuleUtils::ParallelFunctionAnalysis<News> analysis(
+      *module, [&](Function* func, News& news) {
+        if (func->imported()) {
+          return;
+        }
+
+        NewFinder finder;
+        finder.walk(func->body);
+        news = std::move(finder.news);
+      });
+
+    // Also find news in the module scope.
+    NewFinder moduleFinder;
+    moduleFinder.walkModuleCode(module);
+
+    // Process all the news to find the ones we want to modify, adding them to
+    // newsToModify. Note that we must do so in a deterministic order.
+    ModuleUtils::iterDefinedFunctions(
+      *module, [&](Function* func) { processNews(analysis.map[func]); });
+    processNews(moduleFinder.news);
+
+    // Modify the ones we found are relevant. We must modify them all at once as
+    // in the isorecursive type system we want to create a single new rec group
+    // for them all (see below).
+    modifyNews();
+  }
+
+  News newsToModify;
+
+  // All the types that are seen in struct.new etc. operations anywhere in the
+  // program. We use this below to check for an error condition regarding rec
+  // groups.
+  std::unordered_set<HeapType> allSeenTypes;
+
+  // As we generate new names, use a consistent index.
+  Index nameCounter = 0;
+
+  void processNews(const News& news) {
+    for (auto* curr : news.structNews) {
+      if (isInteresting(curr)) {
+        newsToModify.structNews.push_back(curr);
+        allSeenTypes.insert(curr->type.getHeapType());
```

Comment:
This may miss types that are used e.g. in casts but not in allocations. We should still error out in those cases, so it would be better to use `ModuleUtils::collectHeapTypes` to find all the types.

Reply:
Good catch! Done.